### ISSUE TITLE
filesystem permissions not backed up correctly

### ISSUE THREAD

**Issue by kzshantonu** [offen/docker-volume-backup#71]
original ownership:
<img width="408" alt="WindowsTerminal_DmHzUgqMJg" src="https://user-images.githubusercontent.com/7745374/154819139-7666a267-83fc-4987-bc4c-936ba13645b0.png">

ownership of extracted backup archive:
<img width="434" alt="WindowsTerminal_YzuVZEF77u" src="https://user-images.githubusercontent.com/7745374/154819147-1ca88db6-05be-4863-9d0f-87b8989884a0.png">

**Comment by m90** [comment]
Could you provide a reproducible example for this issue? This does not happen for the setups I use this image for so I can't really tell what is happening here judging from two screenshots. It would also be helpful to provide the information requested in the issue template https://github.com/offen/docker-volume-backup/blob/main/.github/ISSUE_TEMPLATE.md

**Comment by m90** [comment]
Is there anything special about the directories for which this happens? Could it be that these are empty for example?

**Comment by kzshantonu** [comment]
It appears that it's the opposite. Dirs with data in them get owned by root whereas empty dirs have original ownership. Container in question is a standard postgres container. Path being backed up is  `/var/lib/postgresql/data` inside container.

* **I'm submitting a ...**
  - [ x] bug report
  - [ ] feature request
  - [ ] support request

* **What is the current behavior?**

original filesystem permissions/ownership not preserved. some dirs get owned by root

* **If the current behavior is a bug, please provide the configuration and steps to reproduce and if possible a minimal demo of the problem.**

* **What is the expected behavior?**

keep original filesystem permissions/ownership

* **What is the motivation / use case for changing the behavior?**

* **Please tell us about your environment:**

  - Image version: 
  - Docker version: 20.10.12
  - docker-compose version: v2.2.3
  - container: postgres:14-alpine

**Comment by kzshantonu** [comment]
I have a reproducible compose yml (trigger manually):

```yaml
version: "3.8"
services:
  db:
    image: postgres:14-alpine
    container_name: postgres_debug_q7Im
    restart: unless-stopped
    labels:
      - docker-volume-backup.stop-during-backup=db
    volumes:
      - postgres_debug_q7Im:/var/lib/postgresql/data
    environment:
      - POSTGRES_PASSWORD=1FHJMSwt0yhIN1zS7I4DilGUhThBKq0x
      - POSTGRES_USER=q7Im
      - POSTGRES_DB=q7Im
  backup:
    image: offen/docker-volume-backup:latest
    restart: always
    environment:
      BACKUP_FILENAME: "backup-%Y-%m-%dT%H-%M-%S.tar.gz"
      BACKUP_FILENAME_EXPAND: "true"
      BACKUP_LATEST_SYMLINK: "backup.latest.tar.gz"
      BACKUP_RETENTION_DAYS: "7"
      BACKUP_PRUNING_PREFIX: "backup-"
      BACKUP_STOP_CONTAINER_LABEL: "db"
    volumes:
      - postgres_debug_q7Im:/backup/my-app-backup:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /root/volbaks_debug:/archive

volumes:
  postgres_debug_q7Im:
    name: postgres_debug_q7Im
```

**Comment by m90** [comment]
Thanks for the example which helped me reproduce your issue. This seems to be an issue in this repo https://github.com/m90/targz

I'll see if I can find out why this is happening there (it's a bit mysterious to me right now because most work is offloaded to the go stdlib there and I would assume someone had noticed that already). I'll let you know what I find out.

**Comment by m90** [comment]
One more question: I also noticed in the backup that it's only the directories which are owned by `root`, all files are owned by UID 70 (which I would guess is Postgres). How exactly do you untar the backup? This could also be something that inadvertently happens on untaring?

**Comment by m90** [comment]
After some digging I think what's happening here is the following: a non-empty directory is not explicitly written to the tar archive explicitly, but instead its contents are, implicitly specifying a location.

When you untar such an archive, the "missing" directories will be created and owned by whoever is running `tar` at that moment, not who has been the owner at archive time. I can confirm this by mounting such a "broken" backup back into a `postgres:14-alpine` container and extracting the archive there, which gives me the expected ownership:

```
/tmp/backup/my-app-backup $ ls -lah
total 132K   
drwx------   19 postgres postgres    4.0K Feb 20 10:15 .
drwxr-xr-x    3 postgres postgres    4.0K Feb 20 10:15 ..
-rw-------    1 postgres postgres       3 Feb 20 10:07 PG_VERSION
drwxr-xr-x    6 postgres postgres    4.0K Feb 20 10:15 base
drwx------    2 postgres postgres    4.0K Feb 20 10:15 global
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_commit_ts
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_dynshmem
-rw-------    1 postgres postgres    4.7K Feb 20 10:07 pg_hba.conf
-rw-------    1 postgres postgres    1.6K Feb 20 10:07 pg_ident.conf
drwxr-xr-x    4 postgres postgres    4.0K Feb 20 10:15 pg_logical
drwxr-xr-x    4 postgres postgres    4.0K Feb 20 10:15 pg_multixact
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_notify
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_replslot
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_serial
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_snapshots
drwx------    2 postgres postgres    4.0K Feb 20 10:15 pg_stat
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_stat_tmp
drwx------    2 postgres postgres    4.0K Feb 20 10:15 pg_subtrans
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_tblspc
drwx------    2 postgres postgres    4.0K Feb 20 10:07 pg_twophase
drwx------    3 postgres postgres    4.0K Feb 20 10:15 pg_wal
drwx------    2 postgres postgres    4.0K Feb 20 10:15 pg_xact
-rw-------    1 postgres postgres      88 Feb 20 10:07 postgresql.auto.conf
-rw-------    1 postgres postgres   28.0K Feb 20 10:07 postgresql.conf
-rw-------    1 postgres postgres      24 Feb 20 10:07 postmaster.opts
```

I need to think about how and where to fix this best a little (and if it can even be fixed), but I guess these findings should give you a workaround for now?

**Comment by m90** [comment]
So I think I fixed the issue in #72 however this means I needed to refactor some vendored code in package `targz` that I never really understood. As this functionality is vital to this image, I would prefer not to merge this directly but do some proper testing beforehand so users can be sure there are no regressions in the created archives, potentially breaking restore procedures.

If you find the time, it would be much appreciated if you could try building the version from the `dir-ownership` branch

```
git clone git@github.com:offen/docker-volume-backup.git
cd docker-volume-backup
git checkout dir-ownership
docker build -t offen/docker-volume-backup:local .
```

and then try to use that image in your setup, checking if the issue persists.

I will also start using this version against some less critical backups of mine so I can observe if there are issues cropping up.

**Comment by kzshantonu** [comment]
Interesting findings. I'll try this branch and let you know in sometime

**Comment by kzshantonu** [comment]
I can confirm that it seems to be working fine now. May I suggest an easier instruction to restore on the README?
It just seems unnecessarily complicated. Also it doesn't tell to remove the volume first, which can be a problem if the volume already exists.

This is what I propose:

`docker volume rm volume_name` # this removes volume if it exists

`docker run --rm -it -v volume_name:/backup/my-app-backup -v /path/to/local_backups:/archive alpine tar -xvzf /archive/full_filename.tar.gz` # this creates volume first

**Comment by m90** [comment]
Thanks for checking. The branch also works for me, so I'll merge it and release it soon.

About your suggestion: I'm not entirely sure why you would need / want to remove the volume? From what I understand you should be able to keep the existing volume and replace its contents with the untared archive?

**Comment by m90** [comment]
The permissions fix is now included in [v2.13.0](https://github.com/offen/docker-volume-backup/releases/tag/v2.13.0)

**Comment by m90** [comment]
I moved discussion about the restore procedure to #75
